src/unifrac_task_noclass.cpp

```diff
@@ -2210,8 +2131,510 @@ void SUCMP_NM::UnifracVawUnnormalizedUnweightedTask<TFloat>::_run(unsigned int f
+		const unsigned int filled_embs,
+		const uint64_t start_idx, const uint64_t stop_idx,
+		const uint64_t n_samples, const uint64_t n_samples_r,
+		const float * const __restrict__ lengths,
+		const uint32_t * const __restrict__ embedded_proportions,
+		const float  * const __restrict__ embedded_counts,
+		const float  * const __restrict__ sample_total_counts,
+		float * const __restrict__ dm_stripes_buf,
+		float * const __restrict__ dm_stripes_total_buf) {
+   run_VawUnweightedTask_T(
+		   filled_embs,
+                   start_idx, stop_idx, n_samples, n_samples_r,
+		   lengths, embedded_proportions, embedded_counts, sample_total_counts,
+		   dm_stripes_buf, dm_stripes_total_buf);
+}
+
+template<>
+void SUCMP_NM::run_VawUnweightedTask(
+		const unsigned int filled_embs,
+		const uint64_t start_idx, const uint64_t stop_idx,
+		const uint64_t n_samples, const uint64_t n_samples_r,
+		const double * const __restrict__ lengths,
+		const uint32_t * const __restrict__ embedded_proportions,
+		const double  * const __restrict__ embedded_counts,
+		const double  * const __restrict__ sample_total_counts,
+		double * const __restrict__ dm_stripes_buf,
+		double * const __restrict__ dm_stripes_total_buf) {
+   run_VawUnweightedTask_T(
+		   filled_embs,
+                   start_idx, stop_idx, n_samples, n_samples_r,
+		   lengths, embedded_proportions, embedded_counts, sample_total_counts,
+		   dm_stripes_buf, dm_stripes_total_buf);
+}
+
+// ==================================
+
+template<>
+void SUCMP_NM::run_VawUnnormalizedUnweightedTask(
+		const unsigned int filled_embs,
+		const uint64_t start_idx, const uint64_t stop_idx,
+		const uint64_t n_samples, const uint64_t n_samples_r,
+		const float * const __restrict__ lengths,
+		const uint32_t * const __restrict__ embedded_proportions,
+		const float  * const __restrict__ embedded_counts,
+		const float  * const __restrict__ sample_total_counts,
+		float * const __restrict__ dm_stripes_buf) {
+   run_VawUnnormalizedUnweightedTask_T(
+		   filled_embs,
+                   start_idx, stop_idx, n_samples, n_samples_r,
+		   lengths, embedded_proportions, embedded_counts, sample_total_counts,
+		   dm_stripes_buf);
+}
 
-   // next iteration will use the alternative space
-   this->set_alt_embedded_proportions();
+template<>
+void SUCMP_NM::run_VawUnnormalizedUnweightedTask(
+		const unsigned int filled_embs,
+		const uint64_t start_idx, const uint64_t stop_idx,
+		const uint64_t n_samples, const uint64_t n_samples_r,
+		const double * const __restrict__ lengths,
+		const uint32_t * const __restrict__ embedded_proportions,
+		const double  * const __restrict__ embedded_counts,
+		const double  * const __restrict__ sample_total_counts,
+		double * const __restrict__ dm_stripes_buf) {
+   run_VawUnnormalizedUnweightedTask_T(
+		   filled_embs,
+                   start_idx, stop_idx, n_samples, n_samples_r,
+		   lengths, embedded_proportions, embedded_counts, sample_total_counts,
+		   dm_stripes_buf);
```

Comment:
Not for consideration in this PR, and unsure whether this would be useful to pursue, but it seems like a lot of content here follows a similar pattern. I wonder whether it would be more maintainable to generate the code?

Reply:
The templating is what "generates the code".
This is just a way to tell the compiler what variant we need.
Unfortunately, I cannot think of a simpler way.

Reply:
I know, but there is also a lot of pattern here. What I was thinking of, and this may be crazy, but jinja2 templates and just generate it with python -- I saw something like that [here](https://github.com/rustyconover/duckdb-datasketches/tree/main/codegen), but again, it may not actually make sense in this repo for many reasons

Reply:
You are right it would make longer-term maintenance easier.
(Even though it would make code harder to read)
Will merge this PR, and do the code generation variant as a follow-up PR.